etl.py

```diff
@@ -0,0 +1,171 @@
+
+        input_data = self.read_csv(input_file)
+
+        input_data = self.extract_headers(input_data)
+
+        with open(columns_mapping_file, 'r') as f:
+            self.columns_mapping = json.load(f)
+
+        transformed_data = self.transform_data(input_data)
+
+        self.output_file_headers = [column for column in self.columns_mapping.values() if column is not None]
+
+        self.write_csv(output_file, transformed_data)
+
+    def extract_headers(self, input_data: List[List[Any]]):
+        """Save the headers before removing the first row and remove headers from data."""
+        self.input_file_headers = input_data[0] if input_data else []
+        if input_data:
+            input_data_without_headers = input_data[1:]
+            return input_data_without_headers
+        else:
+            return input_data
+
+    def read_csv(self, filename: str) -> List[List[Any]]:
+        """Reads a CSV file and returns the data as a list of rows."""
+        data: List[List[Any]] = []
+        try:
+            with open(filename, 'r', newline='') as file:
+                reader = csv.reader(file)
+                for row in reader:
+                    data.append(row)
+        except FileNotFoundError:
+            print(f"Error: Could not find the file '{filename}'")
+        except Exception as e:
+            print(f"Error while reading the file '{filename}': {e}")
+        
+        return data
+
+    def write_csv(self, filename: str, data: List[List[Any]]) -> None:
+        """Writes the data to a CSV file with the provided headers."""
+        try:
+            with open(filename, 'w', newline='') as file:
+                writer = csv.writer(file)
+                writer.writerow(self.output_file_headers)
+                writer.writerows(data)
+            print(f"The formatted data has been written to '{filename}'")
+        except Exception as e:
+            print(f"Error writing to the file '{filename}': {e}")
+
+    def transform_data(
+            self, 
+            input_data: List[List[Any]], 
+        ) -> List[List[Any]]:
+        """Transforms the data according to the example headers."""
+        transformed_data: List[List[Any]] = []
+
+        for input_row in input_data:
+            transformed_row = []
+            for column_name in self.input_file_headers:
+                
+                # Get the input column name corresponding to the output column name
+                input_column_index = list(self.columns_mapping.keys()).index(column_name)
+
+                if list(self.columns_mapping.values())[input_column_index] is None:
+                    continue
+
+                # Get the value of the input column
+                value = input_row[input_column_index]
+
+                transformed_value = self.transform_column_value(column_name, value)
+                transformed_row.append(transformed_value)
+            transformed_data.append(transformed_row)
```

Comment:
Again, this could be a generator using `yield` to avoid building out a large intermediate data structure